# PROJECT 3 CNN

## Part 1 : Importing modules

In [17]:
import numpy as np
import tensorflow as tf
from tensorflow.keras.utils import to_categorical
import cv2
import os
import matplotlib.pyplot as plt
from random import shuffle

## Part 2 : Setting the hyperparameters

In [19]:
IMAGE_SIZE = 128
FC_SIZE = 32

## Part 3 : Read image data set

In [20]:
DATA_DIR = os.path.join(os.getcwd() , 'datasets' , 'p_3')

TRAIN_DIR = os.path.join(DATA_DIR , 'train')
TEST_DIR = os.path.join(DATA_DIR , 'test')
AUG_DIR = os.path.join(DATA_DIR , 'aug')

categories = ['cats' , 'dogs']

In [21]:
training_data = []
testing_data = []

def create_training_data():
    for category in categories :
        path = os.path.join(TRAIN_DIR , category)
        category_index = categories.index(category)
        for image in os.listdir(path):
            image_array = cv2.imread(os.path.join(path , image) )
            resized_array = cv2.resize(image_array , (IMAGE_SIZE , IMAGE_SIZE ))
            training_data.append([resized_array , category_index])
def create_testing_data():
    for category in categories :
        path = os.path.join(TEST_DIR , category)
        category_index = categories.index(category)
        for image in os.listdir(path):
            image_array = cv2.imread(os.path.join(path , image) )
            resized_array = cv2.resize(image_array , (IMAGE_SIZE , IMAGE_SIZE ))
            testing_data.append([resized_array , category_index])

create_training_data()
create_testing_data()

shuffle(training_data)
shuffle(testing_data)

In [22]:
x_train = []
y_train = []

for features , label in training_data :
    x_train.append(features)
    y_train.append(label)
    
x_test = []
y_test = []

for features , label in testing_data :
    x_test.append(features)
    y_test.append(label)
    
x_train = np.array(x_train)
x_test = np.array(x_test)

In [23]:
x_train.shape

(40, 128, 128, 3)

## Part 4 : Run TensorFlow model

In [24]:
model = tf.keras.models.Sequential()

model.add(tf.keras.layers.BatchNormalization(input_shape=(IMAGE_SIZE , IMAGE_SIZE , 3 , )))
model.add(tf.keras.layers.Conv2D(32 , kernel_size=(5,5) , activation='relu'))
model.add(tf.keras.layers.MaxPooling2D(pool_size=(2,2) , strides=2))
model.add(tf.keras.layers.Conv2D(64 , kernel_size=(5,5) , activation='relu'))
model.add(tf.keras.layers.MaxPooling2D(pool_size=(2,2) , strides=2))

model.add(tf.keras.layers.Flatten())
model.add(tf.keras.layers.Dense(FC_SIZE , activation='relu'))
model.add(tf.keras.layers.Dropout(0.4))
model.add(tf.keras.layers.Dense(2 , activation='softmax'))

In [25]:
model.summary()

Model: "sequential_1"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
batch_normalization_1 (Batch (None, 128, 128, 3)       12        
_________________________________________________________________
conv2d_2 (Conv2D)            (None, 124, 124, 32)      2432      
_________________________________________________________________
max_pooling2d_2 (MaxPooling2 (None, 62, 62, 32)        0         
_________________________________________________________________
conv2d_3 (Conv2D)            (None, 58, 58, 64)        51264     
_________________________________________________________________
max_pooling2d_3 (MaxPooling2 (None, 29, 29, 64)        0         
_________________________________________________________________
flatten_1 (Flatten)          (None, 53824)             0         
_________________________________________________________________
dense_2 (Dense)              (None, 32)               

In [26]:
model.compile(optimizer='adam' , loss='categorical_crossentropy' , metrics=['accuracy'])

In [27]:
# Fitting without image augmentation
model.fit(x_train , to_categorical(y_train) , epochs=100 , validation_data=(x_test , to_categorical(y_test)))

Train on 40 samples, validate on 20 samples
Epoch 1/100
40/40 [==============================] - 4s 107ms/sample - loss: 1.2931 - accuracy: 0.5750 - val_loss: 15.6053 - val_accuracy: 0.5000
Epoch 2/100
40/40 [==============================] - 2s 39ms/sample - loss: 1.1464 - accuracy: 0.6000 - val_loss: 0.7933 - val_accuracy: 0.6000
Epoch 3/100
40/40 [==============================] - 2s 40ms/sample - loss: 0.9341 - accuracy: 0.6250 - val_loss: 0.7527 - val_accuracy: 0.4500
Epoch 4/100
40/40 [==============================] - 2s 41ms/sample - loss: 0.6183 - accuracy: 0.6750 - val_loss: 1.4535 - val_accuracy: 0.5000
Epoch 5/100
40/40 [==============================] - 2s 41ms/sample - loss: 0.5713 - accuracy: 0.7000 - val_loss: 0.6608 - val_accuracy: 0.6000
Epoch 6/100
40/40 [==============================] - 2s 41ms/sample - loss: 0.4742 - accuracy: 0.8250 - val_loss: 1.1801 - val_accuracy: 0.5000
Epoch 7/100
40/40 [==============================] - 2s 41ms/sample - loss: 0.4286 - accur

Epoch 57/100
40/40 [==============================] - 1s 36ms/sample - loss: 0.0022 - accuracy: 1.0000 - val_loss: 1.1498 - val_accuracy: 0.7500
Epoch 58/100
40/40 [==============================] - 2s 38ms/sample - loss: 0.0013 - accuracy: 1.0000 - val_loss: 1.1654 - val_accuracy: 0.7500
Epoch 59/100
40/40 [==============================] - 1s 22ms/sample - loss: 0.0069 - accuracy: 1.0000 - val_loss: 1.0878 - val_accuracy: 0.7500
Epoch 60/100
40/40 [==============================] - 1s 22ms/sample - loss: 0.0022 - accuracy: 1.0000 - val_loss: 1.0437 - val_accuracy: 0.6500
Epoch 61/100
40/40 [==============================] - 1s 25ms/sample - loss: 0.0011 - accuracy: 1.0000 - val_loss: 1.0852 - val_accuracy: 0.7000
Epoch 62/100
40/40 [==============================] - 1s 31ms/sample - loss: 0.0021 - accuracy: 1.0000 - val_loss: 1.1383 - val_accuracy: 0.7000
Epoch 63/100
40/40 [==============================] - 1s 27ms/sample - loss: 0.0218 - accuracy: 1.0000 - val_loss: 1.0980 - val_ac

In [28]:
# Fitting without image augmentation
model.fit(x_train , to_categorical(y_train) , epochs=200 , validation_data=(x_test , to_categorical(y_test)))

Train on 40 samples, validate on 20 samples
Epoch 1/200
40/40 [==============================] - 1s 24ms/sample - loss: 3.9176e-04 - accuracy: 1.0000 - val_loss: 1.5871 - val_accuracy: 0.8000
Epoch 2/200
40/40 [==============================] - 1s 24ms/sample - loss: 9.9869e-04 - accuracy: 1.0000 - val_loss: 1.6002 - val_accuracy: 0.8000
Epoch 3/200
40/40 [==============================] - 1s 24ms/sample - loss: 0.0016 - accuracy: 1.0000 - val_loss: 1.6066 - val_accuracy: 0.8000
Epoch 4/200
40/40 [==============================] - 1s 25ms/sample - loss: 9.4686e-04 - accuracy: 1.0000 - val_loss: 1.6128 - val_accuracy: 0.8000
Epoch 5/200
40/40 [==============================] - 1s 24ms/sample - loss: 2.9073e-04 - accuracy: 1.0000 - val_loss: 1.6203 - val_accuracy: 0.8000
Epoch 6/200
40/40 [==============================] - 1s 24ms/sample - loss: 1.6039e-04 - accuracy: 1.0000 - val_loss: 1.6282 - val_accuracy: 0.8000
Epoch 7/200
40/40 [==============================] - 1s 24ms/sample - lo

Epoch 56/200
40/40 [==============================] - 1s 21ms/sample - loss: 9.4788e-06 - accuracy: 1.0000 - val_loss: 2.0434 - val_accuracy: 0.7500
Epoch 57/200
40/40 [==============================] - 1s 35ms/sample - loss: 4.6664e-05 - accuracy: 1.0000 - val_loss: 2.0899 - val_accuracy: 0.7500
Epoch 58/200
40/40 [==============================] - 1s 36ms/sample - loss: 3.8384e-05 - accuracy: 1.0000 - val_loss: 2.1398 - val_accuracy: 0.7500
Epoch 59/200
40/40 [==============================] - 1s 25ms/sample - loss: 0.0017 - accuracy: 1.0000 - val_loss: 2.0442 - val_accuracy: 0.7500
Epoch 60/200
40/40 [==============================] - 1s 20ms/sample - loss: 3.3372e-05 - accuracy: 1.0000 - val_loss: 1.9065 - val_accuracy: 0.7500
Epoch 61/200
40/40 [==============================] - 1s 20ms/sample - loss: 1.5219e-05 - accuracy: 1.0000 - val_loss: 1.9034 - val_accuracy: 0.7000
Epoch 62/200
40/40 [==============================] - 1s 32ms/sample - loss: 5.7189e-06 - accuracy: 1.0000 - v

40/40 [==============================] - 1s 19ms/sample - loss: 4.3764e-05 - accuracy: 1.0000 - val_loss: 2.4711 - val_accuracy: 0.6500
Epoch 112/200
40/40 [==============================] - 1s 34ms/sample - loss: 1.8197e-05 - accuracy: 1.0000 - val_loss: 2.4961 - val_accuracy: 0.6500
Epoch 113/200
40/40 [==============================] - 1s 36ms/sample - loss: 3.5101e-04 - accuracy: 1.0000 - val_loss: 2.5303 - val_accuracy: 0.6500
Epoch 114/200
40/40 [==============================] - 1s 28ms/sample - loss: 1.4621e-05 - accuracy: 1.0000 - val_loss: 2.5603 - val_accuracy: 0.6500
Epoch 115/200
40/40 [==============================] - 1s 21ms/sample - loss: 1.7202e-05 - accuracy: 1.0000 - val_loss: 2.5847 - val_accuracy: 0.6500
Epoch 116/200
40/40 [==============================] - 1s 37ms/sample - loss: 2.2935e-05 - accuracy: 1.0000 - val_loss: 2.6051 - val_accuracy: 0.6500
Epoch 117/200
40/40 [==============================] - 1s 36ms/sample - loss: 1.4746e-05 - accuracy: 1.0000 - val_

40/40 [==============================] - 1s 25ms/sample - loss: 1.2381e-04 - accuracy: 1.0000 - val_loss: 2.0981 - val_accuracy: 0.7000
Epoch 167/200
40/40 [==============================] - 1s 24ms/sample - loss: 0.0065 - accuracy: 1.0000 - val_loss: 1.7362 - val_accuracy: 0.6500
Epoch 168/200
40/40 [==============================] - 1s 36ms/sample - loss: 1.6898e-05 - accuracy: 1.0000 - val_loss: 2.3975 - val_accuracy: 0.6000
Epoch 169/200
40/40 [==============================] - 1s 36ms/sample - loss: 0.1310 - accuracy: 0.9250 - val_loss: 2.5597 - val_accuracy: 0.7500
Epoch 170/200
40/40 [==============================] - 1s 35ms/sample - loss: 1.0294e-04 - accuracy: 1.0000 - val_loss: 6.7863 - val_accuracy: 0.6000
Epoch 171/200
40/40 [==============================] - 1s 20ms/sample - loss: 0.9020 - accuracy: 0.8250 - val_loss: 4.8093 - val_accuracy: 0.5500
Epoch 172/200
40/40 [==============================] - 1s 30ms/sample - loss: 0.3692 - accuracy: 0.8750 - val_loss: 2.5167 - v

In [29]:
# Fitting without image augmentation
model.fit(x_train , to_categorical(y_train) , epochs=300 , validation_data=(x_test , to_categorical(y_test)))

Train on 40 samples, validate on 20 samples
Epoch 1/300
40/40 [==============================] - 1s 20ms/sample - loss: 0.0075 - accuracy: 1.0000 - val_loss: 1.7591 - val_accuracy: 0.5500
Epoch 2/300
40/40 [==============================] - 1s 29ms/sample - loss: 0.0065 - accuracy: 1.0000 - val_loss: 1.8007 - val_accuracy: 0.5500
Epoch 3/300
40/40 [==============================] - 1s 36ms/sample - loss: 3.6574e-04 - accuracy: 1.0000 - val_loss: 1.8587 - val_accuracy: 0.5500
Epoch 4/300
40/40 [==============================] - 1s 33ms/sample - loss: 1.2381e-04 - accuracy: 1.0000 - val_loss: 1.9126 - val_accuracy: 0.5500
Epoch 5/300
40/40 [==============================] - 1s 20ms/sample - loss: 5.0395e-04 - accuracy: 1.0000 - val_loss: 1.9574 - val_accuracy: 0.6000
Epoch 6/300
40/40 [==============================] - 1s 31ms/sample - loss: 2.7209e-04 - accuracy: 1.0000 - val_loss: 1.9941 - val_accuracy: 0.6000
Epoch 7/300
40/40 [==============================] - 1s 37ms/sample - loss: 

40/40 [==============================] - 1s 37ms/sample - loss: 1.5583e-05 - accuracy: 1.0000 - val_loss: 2.3608 - val_accuracy: 0.6000
Epoch 57/300
40/40 [==============================] - 1s 37ms/sample - loss: 4.2523e-04 - accuracy: 1.0000 - val_loss: 2.3646 - val_accuracy: 0.6000
Epoch 58/300
40/40 [==============================] - 1s 36ms/sample - loss: 3.4025e-05 - accuracy: 1.0000 - val_loss: 2.3669 - val_accuracy: 0.6000
Epoch 59/300
40/40 [==============================] - 1s 21ms/sample - loss: 0.0021 - accuracy: 1.0000 - val_loss: 2.3729 - val_accuracy: 0.6000
Epoch 60/300
40/40 [==============================] - 1s 28ms/sample - loss: 6.4357e-05 - accuracy: 1.0000 - val_loss: 2.3803 - val_accuracy: 0.6000
Epoch 61/300
40/40 [==============================] - 1s 36ms/sample - loss: 9.9707e-06 - accuracy: 1.0000 - val_loss: 2.3847 - val_accuracy: 0.6000
Epoch 62/300
40/40 [==============================] - 1s 36ms/sample - loss: 6.7324e-05 - accuracy: 1.0000 - val_loss: 2.39

40/40 [==============================] - 1s 37ms/sample - loss: 0.0014 - accuracy: 1.0000 - val_loss: 2.5219 - val_accuracy: 0.6000
Epoch 112/300
40/40 [==============================] - 1s 36ms/sample - loss: 7.7719e-04 - accuracy: 1.0000 - val_loss: 2.5293 - val_accuracy: 0.6000
Epoch 113/300
40/40 [==============================] - 1s 24ms/sample - loss: 6.7913e-05 - accuracy: 1.0000 - val_loss: 2.5366 - val_accuracy: 0.6000
Epoch 114/300
40/40 [==============================] - 1s 24ms/sample - loss: 0.0176 - accuracy: 0.9750 - val_loss: 2.5418 - val_accuracy: 0.6000
Epoch 115/300
40/40 [==============================] - 1s 35ms/sample - loss: 1.7893e-05 - accuracy: 1.0000 - val_loss: 2.5458 - val_accuracy: 0.6000
Epoch 116/300
40/40 [==============================] - 1s 36ms/sample - loss: 1.0618e-04 - accuracy: 1.0000 - val_loss: 2.5492 - val_accuracy: 0.6000
Epoch 117/300
40/40 [==============================] - 1s 36ms/sample - loss: 1.2855e-05 - accuracy: 1.0000 - val_loss: 2.

Epoch 166/300
40/40 [==============================] - 2s 41ms/sample - loss: 2.9576e-05 - accuracy: 1.0000 - val_loss: 2.8070 - val_accuracy: 0.6000
Epoch 167/300
40/40 [==============================] - 2s 40ms/sample - loss: 4.3480e-06 - accuracy: 1.0000 - val_loss: 2.8075 - val_accuracy: 0.6000
Epoch 168/300
40/40 [==============================] - 2s 40ms/sample - loss: 2.1904e-06 - accuracy: 1.0000 - val_loss: 2.8117 - val_accuracy: 0.6000
Epoch 169/300
40/40 [==============================] - 2s 41ms/sample - loss: 1.9169e-05 - accuracy: 1.0000 - val_loss: 2.8147 - val_accuracy: 0.6000
Epoch 170/300
40/40 [==============================] - 2s 41ms/sample - loss: 2.6971e-06 - accuracy: 1.0000 - val_loss: 2.8173 - val_accuracy: 0.6000
Epoch 171/300
40/40 [==============================] - 2s 41ms/sample - loss: 1.7943e-05 - accuracy: 1.0000 - val_loss: 2.8183 - val_accuracy: 0.6000
Epoch 172/300
40/40 [==============================] - 2s 40ms/sample - loss: 4.3799e-05 - accuracy:

Epoch 221/300
40/40 [==============================] - 2s 39ms/sample - loss: 3.1774e-05 - accuracy: 1.0000 - val_loss: 3.1468 - val_accuracy: 0.6000
Epoch 222/300
40/40 [==============================] - 2s 41ms/sample - loss: 1.8000e-06 - accuracy: 1.0000 - val_loss: 3.1577 - val_accuracy: 0.6000
Epoch 223/300
40/40 [==============================] - 2s 41ms/sample - loss: 1.5795e-07 - accuracy: 1.0000 - val_loss: 3.1652 - val_accuracy: 0.6000
Epoch 224/300
40/40 [==============================] - 2s 40ms/sample - loss: 4.4081e-05 - accuracy: 1.0000 - val_loss: 3.1710 - val_accuracy: 0.6000
Epoch 225/300
40/40 [==============================] - 2s 41ms/sample - loss: 5.6555e-04 - accuracy: 1.0000 - val_loss: 3.1773 - val_accuracy: 0.6000
Epoch 226/300
40/40 [==============================] - 2s 40ms/sample - loss: 7.2419e-07 - accuracy: 1.0000 - val_loss: 3.1828 - val_accuracy: 0.6000
Epoch 227/300
40/40 [==============================] - 2s 41ms/sample - loss: 1.4334e-04 - accuracy:

Epoch 276/300
40/40 [==============================] - 2s 41ms/sample - loss: 1.3369e-04 - accuracy: 1.0000 - val_loss: 4.6592 - val_accuracy: 0.5500
Epoch 277/300
40/40 [==============================] - 2s 41ms/sample - loss: 0.0012 - accuracy: 1.0000 - val_loss: 4.5937 - val_accuracy: 0.5500
Epoch 278/300
40/40 [==============================] - 2s 41ms/sample - loss: 7.9939e-05 - accuracy: 1.0000 - val_loss: 4.5379 - val_accuracy: 0.5500
Epoch 279/300
40/40 [==============================] - 2s 40ms/sample - loss: 1.3828e-06 - accuracy: 1.0000 - val_loss: 4.4966 - val_accuracy: 0.5500
Epoch 280/300
40/40 [==============================] - 2s 41ms/sample - loss: 1.5501e-04 - accuracy: 1.0000 - val_loss: 4.4485 - val_accuracy: 0.5500
Epoch 281/300
40/40 [==============================] - 2s 41ms/sample - loss: 1.6613e-05 - accuracy: 1.0000 - val_loss: 4.4100 - val_accuracy: 0.5500
Epoch 282/300
40/40 [==============================] - 2s 40ms/sample - loss: 1.3399e-05 - accuracy: 1.0

### Report

* For model that is trained with **100** epochs : **training_acc = 100 % , testing_acc = 80%**
* For model that is trained with **200** epochs : **training_acc = 100 % , testing_acc = 55%**
* For model that is trained with **300** epochs : **training_acc = 100 % , testing_acc = 60%**